docs/source/quadrature.rst

```diff
@@ -0,0 +1,113 @@
+.. only:: html
+
+  .. contents::
+
+Quadrature rules
+================
+
+To numerically compute the integrals in the variational formulation,
+a quadrature rule is required.
+By default, Firedrake obtains a quadrature rule by estimating the polynomial
+degree of the integrands within a :py:class:`ufl.Form`. Sometimes
+this estimate might be quite large, and a warning like this one will be raised:
+
+.. code-block::
+
+   tsfc:WARNING Estimated quadrature degree 13 more than tenfold greater than any argument/coefficient degree (max 1)
+
+Sometimes the estimated quadrature degree might be even in the hundreds or thousands,
+rendering the integration prohibitively expensive. 
+
+Specifying the quadrature rule in the variational formulation
+-------------------------------------------------------------
+
+To manually override the default, the
+quadrature degree can be prescribed on each integral :py:class:`~ufl.measure.Measure`,
+
+.. code-block:: python3
+
+   inner(sin(u)**4, v) * dx(degree=4)
+
+Setting ``degree=4`` means that the quadrature rule will be exact only for integrands
+of total polynomial degree up to 4. This, of course, will introduce a greater numerical error than the default.
+
+The default quadrature degree estimation may be overridden by setting the ``"quadrature_degree"`` in
+the ``form_compiler_parameters`` dictionary passed on to
+:py:func:`~.solve`, :py:func:`~.project`, or :py:class:`~.NonlinearVariationalProblem`.
+
+.. code-block:: python3
+
+   F = inner(grad(u), grad(v))*dx(degree=0) + inner(exp(u), v)*dx - inner(1, v)*dx
+
+   solve(F == 0, u, form_compiler_parameters={"quadrature_degree": 4})
+
+In the example above, only the integrals with unspecified quadrature degree
+will be computed on a quadrature rule that exactly integrates polynomials of
+the degree set in ``form_compiler_parameters``.
+
+Another way to specify the quadrature rule is through the ``scheme`` keyword. This could be
+either a :py:class:`~finat.quadrature.QuadratureRule`, or a string. Supported string values
+are ``"default"``, ``"canonical"``, ``"lump"``, and ``"KMV"``. For more details see
+:py:func:`~FIAT.quadrature_schemes.create_quadrature`.  
+
+Lumped quadrature schemes
+-------------------------
+
+Spectral elements, such as Gauss-Legendre-Lobatto and KMV, may be used with
+lumped quadrature schemes to produce a diagonal mass matrix.
+
+.. code-block:: python3
+
+   degree = 3
+   V = FunctionSpace(mesh, "KMV", degree)
+   u = TrialFunction(V)
+   v = TestFunction(V)
+
+   inner(u, v) * dx(scheme="lump", degree=degree)
```

Comment:
In general we try to literal-include any code so that it won't bit rot (e.g. see the interpolation docs). Any chance you could do that?